architecture/src/androidLibMain/kotlin/navigation/Navigator.kt

```diff
@@ -138,24 +141,45 @@ class ActivityNavigator<A : NavigationAction<*>>(private val navigationMapper: (
     }
 
     private fun removeFragment(removeFragmentSpec: NavigationSpec.RemoveFragment) {
-        assert(manager?.fragmentManager != null)
-        val fragmentManager = manager?.fragmentManager ?: return
+        assert(manager != null)
+        val fragmentManager = removeFragmentSpec.getFragmentManager(manager!!)
         val fragment = fragmentManager.findFragmentByTag(removeFragmentSpec.tag) ?: return
 
         val transaction = fragmentManager.beginTransaction()
         transaction.remove(fragment)
         transaction.commit()
     }
 
+    private fun popFragment(popFragmentSpec: NavigationSpec.PopFragment) {
+        assert(manager != null)
+        val fragmentManager = popFragmentSpec.getFragmentManager(manager!!)
+        if (popFragmentSpec.immediate) {
+            fragmentManager.popBackStackImmediate()
+        } else {
+            fragmentManager.popBackStack()
+        }
+    }
+
+    private fun popFragmentTo(popToFragmentSpec: NavigationSpec.PopFragmentTo) {
+        assert(manager != null)
+        val fragmentManager = popToFragmentSpec.getFragmentManager(manager!!)
```

Comment:
It wont get smartcasted for an assert in any case.
Refactoring navigation to become either suspended or return an error state is possible of course, but it would be a pretty significan api update that has impact on nearly all ViewModels. So id definitely not do it here. Furthermore I think its actually good to crash. Navigation shouldnt happen if a ViewModel is not in the lifecycle and this reflects that.